### PR TITLE
Generation resizing

### PR DESCRIPTION
**Pull request by earthling-amzn**
These changes have the generational mode track the minimum mutator utilization (percentage of process time used by mutators). When it falls below a configuration percentage (GCTimeRatio), a heuristic will transfer memory capacity to whatever generation has been using more CPU time. The assumption here is that by increasing capacity, we will decrease the collection frequency and improve the MMU.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Reviewers
 * [Roman Kennke](https://openjdk.org/census#rkennke) (@rkennke - **Reviewer**) ⚠️ Review applies to [d7a01946](https://git.openjdk.org/shenandoah/pull/177/files/d7a0194688ce83d7faada2e847cfe18ce8aafc14)
 * [Kelvin Nilsen](https://openjdk.org/census#kdnilsen) (@kdnilsen - Committer) ⚠️ Review applies to [41f057fa](https://git.openjdk.org/shenandoah/pull/177/files/41f057facc9a3e65e54da004b8ba9aa9e861b83b)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/shenandoah pull/177/head:pull/177` \
`$ git checkout pull/177`

Update a local copy of the PR: \
`$ git checkout pull/177` \
`$ git pull https://git.openjdk.org/shenandoah pull/177/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 177`

View PR using the GUI difftool: \
`$ git pr show -t 177`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/shenandoah/pull/177.diff">https://git.openjdk.org/shenandoah/pull/177.diff</a>

</details>

### CODE REVIEW

**Comment by rkennke** [comment]
Thanks, William! The PR has merge conflicts, can you resolve them?

Thanks,
Roman

File: src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp
Comment (rkennke) [inline src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp:451]: Why does the GCIdMark need to move around?

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp
Comment (rkennke) [inline src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:1791]: What is that logging for/ what does the log message mean? I'd either improve the log message or remove the logging (or make it dev+trace) if it was only for dev purposes.

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp
Comment (rkennke) [inline src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:1804]: Same here.

File: src/hotspot/share/gc/shenandoah/shenandoahMmuTracker.cpp
Comment (rkennke) [inline src/hotspot/share/gc/shenandoah/shenandoahMmuTracker.cpp:25]: You need to include precompiled.hpp here.

File: src/hotspot/share/gc/shenandoah/shenandoahMmuTracker.hpp
Comment (rkennke) [inline src/hotspot/share/gc/shenandoah/shenandoahMmuTracker.hpp:28]: What do we need the iterator.hpp for?

**Review by rkennke — COMMENTED** [review]
Thank you for implementing this useful change! I have a few questions and comments.

File: src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp:451]: I pulled up the gcid mark because every old generation collection is preceded by a "bootstrap" young collection. Logically, this bootstrap phase belongs to the old collection so it should have the same GC id as the subsequent old marking phase.

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:1804]: Will remove this.

File: src/hotspot/share/gc/shenandoah/shenandoahMmuTracker.cpp
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahMmuTracker.cpp:25]: Done (added this to my new file template as well).

**Review by kdnilsen — APPROVED** [review]
LGTM.  I'm not yet convinced that this is the right heuristic, or the only heuristic for resizing generations.  But this is a huge step making it possible to adjust generation sizes on the fly.  I expect further refinement will be driven by additional experiments.

**Review by rkennke — APPROVED** [review]
Looks good to me. Thank you!

EDIT: well actually, it is still complaining about conflict, and you should also make it ready for review ;-)

**Comment by ysramakrishna** [comment]
@earthling-amzn : I have a partially completed review from last night which I'll complete and post here in the next hour or so. Sorry for the delay.

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp
Comment (ysramakrishna) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:1098]: Left a related comment in `MmuTracker::adjust_generational_size()`.

File: src/hotspot/share/gc/shenandoah/shenandoahMmuTracker.hpp
Comment (ysramakrishna) [reply src/hotspot/share/gc/shenandoah/shenandoahMmuTracker.hpp:37]: I'd mention the interval `GCPauseIntervalMillis` here for clarity. (I'd say it's a curious naming of the interval, but it's already used in that sense, so we leave it as is.)

File: src/hotspot/share/gc/shenandoah/shenandoahMmuTracker.hpp
Comment (ysramakrishna) [reply src/hotspot/share/gc/shenandoah/shenandoahMmuTracker.hpp:90]: I think the way you use it, it's not 20% of the heap but rather 20% of the free space in the generation that will provide the transfer delta. May be reword for clarity?

**Review by ysramakrishna — COMMENTED** [review]
Sorry again for not getting this review back to you in time.

It looks good overall, but some comments here for you to use to perhaps improve a few things.

Reviewed & approved, modulo the above comments.

File: src/hotspot/share/gc/shenandoah/shenandoahMmuTracker.hpp
Comment (ysramakrishna) [reply src/hotspot/share/gc/shenandoah/shenandoahMmuTracker.hpp:44]: I see now that it's a decaying average over 5 second intervals. May be elaborate the comment accrdingly.

File: src/hotspot/share/gc/shenandoah/shenandoahMmuTracker.cpp
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahMmuTracker.cpp:146]: Yes, this method is a bit long. The similarity breaks a bit because it needs to enforce the min or max constraint depending on the direction of the transfer. I'll split it up in a subsequent PR.

File: src/hotspot/share/gc/shenandoah/shenandoahMmuTracker.hpp
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahMmuTracker.hpp:54]: It's vestigial - I was using it originally to "verify" the result of the per-generation based MMU. I'll rename it.

File: src/hotspot/share/gc/shenandoah/shenandoahYoungGeneration.cpp
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahYoungGeneration.cpp:95]: I'll turn this into an `is_bootstrapping()` method.

File: src/hotspot/share/gc/shenandoah/shenandoahMmuTracker.cpp
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahMmuTracker.cpp:124]: Good point, I'll make that change.

File: src/hotspot/share/gc/shenandoah/shenandoahMmuTracker.hpp
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahMmuTracker.hpp:90]: Yes, it's 20% of the available capacity.

File: src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp:992]: I will add asserts here.

File: src/hotspot/share/gc/shenandoah/mode/shenandoahGenerationalMode.cpp
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/mode/shenandoahGenerationalMode.cpp:39]: I think it is a confusingly named parameter, but I'm interpreting it based on the description: 
> "Adaptive size policy application time to GC time ratio"

Any time the average MMU drops below this number, it attempts to resize the generations.

File: src/hotspot/share/gc/shenandoah/shenandoahMmuTracker.cpp
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahMmuTracker.cpp:96]: I prefer 'elapsed' here when dealing with time deltas.

File: src/hotspot/share/gc/shenandoah/shenandoahMmuTracker.cpp
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahMmuTracker.cpp:99]: It's vestigial, I'll change it.

**Pull request closed without merging**
